File: leetcode/e/Reverse Words in a String II.java
/**
 * Given an input string, reverse the string word by word. A word is defined as a sequence of non-space characters.
 * <p>
 * The input string does not contain leading or trailing spaces and the words are always separated by a single space.
 * <p>
 * For example,
 * Given s = "the sky is blue",
 * return "blue is sky the".
 **/

public class Solution {
  // swap two ends for reverse, reverse entire string, then reverse each word
  public void reverseWords(char[] s) {
    reverse(s, 0, s.length - 1);
    int lo = 0, hi = 0;
    while (hi <= s.length) {
      if (hi == s.length) reverse(s, lo, hi++ - 1); // last word
      else {
        if (s[hi] == ' ') {
          reverse(s, lo, hi - 1); // reverse each word
          lo = hi + 1;
        }
        hi++;
      }
    }
  }

  private void reverse(char[] s, int lo, int hi) {
    while (lo < hi) {
      char c = s[lo];
      s[lo] = s[hi];
      s[hi] = c;
      lo++;
      hi--;
    }
  }
}

File: lc_plusplus/Reverse Words in a String II.cc
// Given an input string, reverse the string word by word. A word is defined as a sequence of non-space characters.
//
// The input string does not contain leading or trailing spaces and the words are always separated by a single space.
//
// For example,
// Given s = "the sky is blue",
// return "blue is sky the".
//
// Could you do it in-place without allocating extra space?
//
// Related problem: Rotate Array


class Solution {
public:
    void reverseWords(vector<char>& str) {
        reverse(str.begin(), str.end());
        for (size_t i = 0, j = 0; i <= str.size(); ++i) {
            if (i == str.size() || str[i] == ' ') {
                reverse(str.begin() + j, str.begin() + i);
                j = i + 1;
            }
        }
    }
};
